holopy/core/utils.py

```diff
@@ -43,22 +44,33 @@
 
 # FIXME this is difficult to test, since it works on the file level
 # perhaps make this capture the output rather than writing it to devnull?
-class SuppressOutput(object):
-    STD_OUT = 1
+class SuppressOutput():
     def __init__(self, suppress_output=True):
         self.suppress_output = suppress_output
+        try:
+            self.std_out = sys.stdout.fileno()
+        except io.UnsupportedOperation:
+            self.stdout_behaves_normally = False  # ie running on travis
+            self.std_out = 1
+        else:
+            self.stdout_behaves_normally = True
+
+    def _redirect_stdout(self, destination_fileno):
+        os.dup2(destination_fileno, self.std_out)
+        if self.stdout_behaves_normally:
+            sys.stdout.close()
```

Comment:
Question, not a comment: Do we want to close `sys.stdout` on exiting the context manager?

Reply:
Yes, that is intended. It will flush and close the existing `sys.stdout`. The next line sets up a new `sys.stdout` as a low level text buffer. I don't completely understand everything that's going on, but it seems to provide a level of redundancy on top of redirecting output with `os.dup` and `os.dup2`, and stops the rare (approx 1%) issues where output either wasn't being suppressed or wasn't being reinstated. Since both were happening sometimes we need the patch on each end. Unfortunately in an inference calculation something that occurs in 1% of calculations ends up being pretty frequent!